feat: 게시판 CRUD 구현 (level 2)

sangcci opened the pull request:
- 프로젝트 환경 설정
- Entity(Domain), Service, Controller 구현
- JPA Repository 구현
- REST API 설계

[[추가 commit]]
- update()에서 save() 삭제
- @Lob 추가
- domain 객체에 @Setter 제거
- boardResponseDto를 list로 할 때 정적 팩토리 메서드를 사용하여 구현

? Dto, Domain 객체에 Getter는 사용해도 문제없는지..?
? 서비스 레이어의 create() 구현하면서 이렇게 Board를 save하고 그 내용을 보여주기 위해 따로 ResponseDto를 만들어도 되는지? 
? BoardResponseDto와 Board 둘이 정확히 어떤 역할(책임?)을 맡아야 하는지?

!! PUT, PATCH 메서드가 지원하지 않는 오류 발생
- 해결방안 1. application.yml 추가 -> X
- 해결방안 2. CORS 허용 메서드 추가 -> X
- SpringBoot 2.7 버전에 문제가 있는것으로 추정중입니다..

-> 해결하였습니다!
<@PathVariable> 어노테이션을 잘못 이해하고 있었습니다...
계속 URL로 요청할 때 Query String으로 값을 보내고 있어서 생긴 오류였고, 일반 값으로 변경하여 해결할 수 있었습니다.

sangcci committed:
init: 초기 프로젝트 환경설정

- 1. version 체크: Java11, SpringBoot 2.7.*
- 2. dependency 추가: lombok, Spring-Web-MVC, Spring-Data-JPA, MySQL Driver
- 3. application.yml 내용 추가: port 설정, DB연동 및 편의성 설정

sangcci committed:
feat: 게시판 CRUD 구현

- Entity(Domain), Service, Controller 구현
- JPA Repository 구현
- REST API 설계

!! PUT, PATCH 메서드가 지원하지 않는 오류 발생
- 해결방안 1. application.yml 추가 -> X
- 해결방안 2. CORS 허용 메서드 추가 -> X
- SpringBoot 2.7 버전에 문제가 있는것으로 생각중

baksakcci commented:
해결하였습니다!
<@PathVariable> 어노테이션을 잘못 이해하고 있었습니다...
계속 URL로 요청할 때 Query String으로 값을 보내고 있어서 생긴 오류였고, 일반 값으로 변경하여 해결할 수 있었습니다.

galmegiz commented on src/main/java/com/example/spring_jscode/service/BoardService.java:
```diff
@@ -0,0 +1,54 @@
+    public Board update(Long id, String title, String content) {
+        Board board = boardRepository.findById(id).orElseThrow(NullPointerException::new);
+        board.setTitle(title);
+        board.setContent(content);
```
- setter사용보다는 별도 update라는 메소드를 작성해보는 게 어떨까요? 어차피 변화는 도메인 내에서 일어나는 거기 때문에 서비스 레이어의 부담이 줄어들고 setter를 열어둠으로서 발생할 수 있는 문제를 해결할 수 있을 것 같습니다.
- 심화미션 관련 내용 : 도메인 모델에 setter메서드 넣지 않기
https://velog.io/@backfox/setter-%EC%93%B0%EC%A7%80-%EB%A7%90%EB%9D%BC%EA%B3%A0%EB%A7%8C-%ED%95%98%EA%B3%A0-%EA%B0%80%EB%B2%84%EB%A6%AC%EB%A9%B4-%EC%96%B4%EB%96%A1%ED%95%B4%EC%9A%94

sangcci replied:
의견 감사합니다! 덕분에 올려주신 포스팅 읽고 리펙토링 진행해보겠습니다!

galmegiz commented on src/main/java/com/example/spring_jscode/service/BoardService.java:
```diff
@@ -0,0 +1,54 @@
+        Board board = boardRepository.findById(id).orElseThrow(NullPointerException::new);
+        board.setTitle(title);
+        board.setContent(content);
+        boardRepository.save(board);
```
제가 알기론  find를 하면 1차 캐시에  board가 저장되기 때문에   별도 save를 하지 않아도 dirty checking에 의해 update 쿼리가 실행될 겁니다.

galmegiz commented on src/main/resources/application.yml:
```diff
@@ -3,7 +3,7 @@ server:
 spring:
   datasource:
-    url: jdbc:mysql://localhost:3306/jscode
+    url: jdbc:mysql://localhost:3306/jscode?useUnicode=true&characterEncoding=UTF-8
```
저도 깜빡해서 올려버렸는데 설정파일에 비밀번호가 포함된 경우가 많으니 git ignore에 항상 추가해주셔야 합니다!

galmegiz commented on src/main/java/com/example/spring_jscode/entity/Board.java:
```diff
@@ -0,0 +1,29 @@
+
+@Entity
+@Getter
+@Setter
```
클래스 상단에 lombok setter를 넣는 경우 전체 필드에 대한 세터 메소드가 생성되어 버립니다. 이 경우 외부에서 수정하면 안 되는 값(예를 들어 id)값에 대한 setter까지 생성되 버립니다. setter를 개별 field에 넣어주시든지 목적에 맞는 메소드(update 메소드)를 구현하는 걸 권장드립니다.

galmegiz commented on src/main/java/com/example/spring_jscode/entity/Board.java:
```diff
@@ -0,0 +1,29 @@
+    @Column(nullable = false)
+    private String title;
+
+    @Column(nullable = false)
```
@Column을 사용하게 되면 기본적으로 mysql은 varchar(255)입니다. content의 경우 글 길이가 길어질 수 있기 때문에 @Lob을 붙여주셔야 합니다.

galmegiz commented on src/main/java/com/example/spring_jscode/service/BoardService.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @Transactional(readOnly = true)
+    public List<BoardResponseDto> findAll() {
```
다른 메소드에서는  Board를 응답값으로 보내고 있는데 이 메소드에서는 Dto리스트를 응답으로 보내고 있습니다. 서로 다른 레이어에서 도메이을 주고받는 건 적절치 않아 보이므로 다른 메소드도 dto를 전달할 수 있게 바꿔보시면 좋을 듯 합니다.

galmegiz commented on src/main/java/com/example/spring_jscode/dto/BoardResponseDto.java:
```diff
@@ -0,0 +1,16 @@
+    private String title;
+    private String content;
+
+    public void changeBoardToDto(String title, String content) {
```
아마 domain -> dto를 염두에 두고 메소드를 만드신 것 같습니다.
메소드명을 from() 또는 fromEntity()로 바꿔보시고 static메소드로 만들어 보시는 것도 추천드립니다.(나중에 Pageable다룰 때도 static이 더 편하더라구요)

관련 심화 미션 : https://tecoble.techcourse.co.kr/post/2020-05-26-static-factory-method/

sangcci committed:
refactor: board refactoring

- update()에서 save() 삭제
- @Lob 추가
- domain 객체에 @Setter 제거 및 대체 메서드 구현
- boardResponseDto.fromEntity 정적 팩토리 메서드 구현: Entity->dto

? Dto, Domain 객체에 Getter는 사용해도 문제없는지..?
? 서비스 레이어의 create() 구현하면서 이렇게 Board를 save하고 그 내용을 보여주기 위해 따로 ResponseDto를 만들어도 되는지?
? BoardResponseDto와 Board 둘이 정확히 어떤 역할(책임?)을 맡아야 하는지?

galmegiz reviewed (approved):
피드백 완료

galmegiz reviewed (approved):
피드백 완료

galmegiz reviewed (approved):
피드백 완료

galmegiz reviewed (approved):
승인
